src/sage/data_structures/pairing_heap.pyx

```diff
@@ -0,0 +1,1501 @@
+# Class PairingHeap_of_n_integers
+# ==============================================================================
+
+cdef class PairingHeap_of_n_integers(PairingHeap_class):
+    r"""
+    Pairing Heap for items in range [0..n - 1].
+
+    EXAMPLES::
+
+        sage: from sage.data_structures.pairing_heap import PairingHeap_of_n_integers
+        sage: P = PairingHeap_of_n_integers(5); P
+        PairingHeap_of_n_integers: capacity 5, size 0
+        sage: P.push(1, 3)
+        sage: P.push(2, 2)
+        sage: P
+        PairingHeap_of_n_integers: capacity 5, size 2
+        sage: P.top()
+        (2, 2)
+        sage: P.decrease(1, 1)
+        sage: P.top()
+        (1, 1)
+        sage: P.pop()
+        sage: P.top()
+        (2, 2)
+
+    TESTS::
+
+        sage: from sage.data_structures.pairing_heap import PairingHeap_of_n_integers
+        sage: P = PairingHeap_of_n_integers(0)
+        Traceback (most recent call last):
+        ...
+        ValueError: the capacity of the heap must be strictly positive
+        sage: P = PairingHeap_of_n_integers(1); P
+        PairingHeap_of_n_integers: capacity 1, size 0
+        sage: P = PairingHeap_of_n_integers(5)
+        sage: P.push(11, 3)
+        Traceback (most recent call last):
+        ...
+        ValueError: item must be in range 0..4
+    """
+    def __init__(self, size_t n):
+        r"""
+        Construct the ``PairingHeap_of_n_integers`` where items are integers
+        from ``0`` to ``n-1``.
+
+        INPUT:
+
+        - ``n`` -- strictly positive integer; the maximum number of items in the heap
+
+        EXAMPLES::
+
+            sage: from sage.data_structures.pairing_heap import PairingHeap_of_n_integers
+            sage: P = PairingHeap_of_n_integers(5); P
+            PairingHeap_of_n_integers: capacity 5, size 0
+            sage: P.push(1, 2)
+            sage: P
+            PairingHeap_of_n_integers: capacity 5, size 1
+            sage: P.push(2, 3)
+            sage: P
+            PairingHeap_of_n_integers: capacity 5, size 2
+            sage: P.pop()
+            sage: P
+            PairingHeap_of_n_integers: capacity 5, size 1
+            sage: P.push(10, 1)
+            Traceback (most recent call last):
+            ...
+            ValueError: item must be in range 0..4
+            sage: P = PairingHeap_of_n_integers(0)
+            Traceback (most recent call last):
+            ...
+            ValueError: the capacity of the heap must be strictly positive
+            sage: P = PairingHeap_of_n_integers(1); P
+            PairingHeap_of_n_integers: capacity 1, size 0
+        """
+        if not n:
+            raise ValueError("the capacity of the heap must be strictly positive")
+        self.name = "PairingHeap_of_n_integers"
+        self.n = n
+        self.root = NULL
+        self.nodes = <PairingHeapNode *>check_allocarray(n, sizeof(PairingHeapNode))
+        bitset_init(self.active, n)
+        bitset_clear(self.active)
```

Comment:
is it needed ? The bitset is allocated using calloc, so it memory will be 0 after `bitset_init`. The doc says that `bitset_init` allocates an "empty bitset", so it seems to me that it is a feature of the bitset that it will be all 0 after init.

Reply:
you are right.